File: sem1/Homework_8/task8.3/list.cpp
#include "list.h"
#include <iostream>

using namespace std;

bool isEmpty(List const &list)
{
    return list.size == 0;
}

bool contains(List const &list, String value)
{
    ListElement* tmp = list.head;
    for (int i = 0; i < list.size; ++i)
    {
        if (areEqual(tmp->value, value))
            return true;
        tmp = tmp->next;
    }
    return false;
}

//false => list already contains the value
bool add(List &list, String newValue)
{
    if (isEmpty(list))
    {
        list.head = new ListElement;
        //cout << "new(" << (int)list.head << ")\n";
        list.head->value = clone(newValue);
        list.size++;
        return true;
    }

    if (contains(list, newValue))
            return false;

    list.size++;
    ListElement* newElement = new ListElement;
    //cout << "new(" << (int)newElement << ")\n";
    newElement->value = clone(newValue);
    newElement->next = list.head;
    list.head = newElement;
    return true;
}

String pop(List &list)
{
    String result;
    if (isEmpty(list))
        return result;

    ListElement* tmpElement = list.head;
    String tmpValue = clone(tmpElement->value);
    list.head = list.head->next;
    clear(tmpElement->value);
    delete tmpElement;
    //cout << "delete(" << (int)tmpElement << ")\n";
    list.size--;
    return tmpValue;
}

void show(List const &list)
{
    ListElement* tmp = list.head;
    for (int i = 0; i < list.size; ++i)
    {
        output(tmp->value);
        std::cout << ' ';
        tmp = tmp->next;
    }
}

void clear(List &list)
{
    ListElement* tmp = list.head;
    while (!isEmpty(list))
    {
        tmp = tmp->next;
        delete[] list.head->value.data;
        //cout << "delete(" << (int)list.head->value.data << ")\n";
        delete list.head;
        //cout << "delete(" << (int)list.head << ")\n";
        list.head = tmp;
        list.size--;
    }
    list.head = nullptr;
}

File: sem2/Hw1.2/List.java
package com.group144.Degiv;

public class List {

    private class ListElement {
        private int value;
        private ListElement next;

        ListElement(int value, ListElement next){
            this.value = value;
            this.next = next;
        }

        public int getValue(){
            return value;
        }

        public ListElement getNext(){
            return next;
        }

        public void setNext(ListElement next){
            this.next = next;
        }
    }

    private ListElement head = new ListElement(0, null);
    private int size = 0;

    public int getSize(){
        return size;
    }

    public boolean isEmpty(){
        return size == 0;
    }

    public void add(int value, int index){
        if (index > size)
            return;

        ListElement tmp = head;
        for (int i = 0; i < index; ++i){
            tmp = tmp.getNext();
        }

        ListElement newElement = new ListElement(value, tmp.next);
        tmp.setNext(newElement);
        size++;
    }

    public int pop(int index){
        if (index >= size)
            return -1;

        ListElement tmp = head;
        for (int i = 0; i < index; ++i){
            tmp = tmp.getNext();
        }

        int result = tmp.getNext().getValue();
        tmp.setNext(tmp.getNext().getNext());
        size--;
        return result;
    }

    public int peek(int index){
        if (index >= size)
            return -1;

        ListElement tmp = head;
        for (int i = 0; i <= index; ++i){
            tmp = tmp.getNext();
        }

        return tmp.getValue();
    }
}
